common/flatpak-run.c

```diff
@@ -2441,33 +2444,62 @@ forward_file (XdpDbusDocuments *documents,
 
   fd_list = g_unix_fd_list_new ();
   fd_id = g_unix_fd_list_append (fd_list, fd, error);
+  if (fstat (fd, &stbuf) == 0 && S_ISDIR (stbuf.st_mode))
+    is_dir = TRUE;
   close (fd);
 
-  if (!xdp_dbus_documents_call_add_sync (documents,
-                                         g_variant_new ("h", fd_id),
-                                         TRUE, /* reuse */
-                                         FALSE, /* not persistent */
-                                         fd_list,
-                                         &doc_id,
-                                         NULL,
-                                         NULL,
-                                         error))
-    {
-      if (error)
-        g_dbus_error_strip_remote_error (*error);
-      return FALSE;
+  if (!is_dir)
+    {
+      if (!xdp_dbus_documents_call_add_sync (documents,
+                                             g_variant_new ("h", fd_id),
+                                             TRUE, /* reuse */
+                                             FALSE, /* not persistent */
+                                             fd_list,
+                                             &doc_id,
+                                             NULL,
+                                             NULL,
+                                             error))
+        {
+          if (error)
+            g_dbus_error_strip_remote_error (*error);
+          return FALSE;
+        }
+      if (!xdp_dbus_documents_call_grant_permissions_sync (documents,
+                                                           doc_id,
+                                                           app_id,
+                                                           perms,
+                                                           NULL,
+                                                           error))
+        {
+          if (error)
+            g_dbus_error_strip_remote_error (*error);
+          return FALSE;
+        }
     }
-
-  if (!xdp_dbus_documents_call_grant_permissions_sync (documents,
-                                                       doc_id,
-                                                       app_id,
-                                                       perms,
-                                                       NULL,
-                                                       error))
+  else
     {
-      if (error)
-        g_dbus_error_strip_remote_error (*error);
-      return FALSE;
+      // Requires version 4 of AddFull interface
+      if (!xdp_dbus_documents_call_add_full_sync (documents,
+                                                  g_variant_new_fixed_array (G_VARIANT_TYPE_HANDLE, &fd_id, 1, sizeof (gint32)),
+                                                  8, /* export-directory */
```

Comment:
Please define the enum for the flags and use that here.

Reply:
I copied `DocumentAddFullFlags` to `flatpak-run-private.h` from xdg-desktop-portal/document-portal/document-enums.h, though not quite sure if this is OK.